File: Examples/VisualBasic/Exchange/ListFoldersFromExchangeServer.vb
Imports System.Collections.Generic
Imports System.Configuration
Imports System.Net
Imports System.Threading
Imports Aspose.Email.Exchange
Imports Aspose.Email.Mail
Imports Aspose.Email.Outlook

'
'This project uses Automatic Package Restore feature of NuGet to resolve Aspose.Email for .NET API reference 
'when the project is build. Please check https://Docs.nuget.org/consume/nuget-faq for more information. 
'If you do not wish to use NuGet, you can manually download Aspose.Email for .NET API from http://www.aspose.com/downloads, 
'install it and then add its reference to this project. For any issues, questions or suggestions 
'please feel free to contact us using http://www.aspose.com/community/forums/default.aspx
'

Namespace Aspose.Email.Examples.VisualBasic.Email.Exchange
	Class ListFoldersFromExchangeServer
		' ExStart:ListFoldersFromExchangeServer
		Public Shared Sub Run()
			Try
				Dim client As IEWSClient = EWSClient.GetEWSClient("https://outlook.office365.com/ews/exchange.asmx", "testUser", "pwd", "domain")
				Console.WriteLine("Downloading all messages from Inbox....")

				Dim mailboxInfo As ExchangeMailboxInfo = client.GetMailboxInfo()
				Console.WriteLine("Mailbox URI: " + mailboxInfo.MailboxUri)
				Dim rootUri As String = client.GetMailboxInfo().RootUri
				' List all the folders from Exchange server
				Dim folderInfoCollection As ExchangeFolderInfoCollection = client.ListSubFolders(rootUri)
				For Each folderInfo As ExchangeFolderInfo In folderInfoCollection
					' Call the recursive method to read messages and get sub-folders
					ListSubFolders(client, folderInfo)
				Next

				Console.WriteLine("All messages downloaded.")
			Catch ex As Exception
				Console.WriteLine(ex.Message)
			End Try
		End Sub
		Private Shared Sub ListSubFolders(client As IEWSClient, folderInfo As ExchangeFolderInfo)
			' Create the folder in disk (same name as on IMAP server)
			Console.WriteLine(folderInfo.DisplayName)
			Try
				' If this folder has sub-folders, call this method recursively to get messages
				Dim folderInfoCollection As ExchangeFolderInfoCollection = client.ListSubFolders(folderInfo.Uri)
				For Each subfolderInfo As ExchangeFolderInfo In folderInfoCollection
					ListSubFolders(client, subfolderInfo)
				Next
			Catch generatedExceptionName As Exception
			End Try
		End Sub
		' ExEnd:ListFoldersFromExchangeServer
	End Class
End Namespace

File: Examples/CSharp/Exchange_EWS/ListFoldersFromExchangeServer.cs
using System;
using System.Collections.Generic;
using System.Configuration;
using System.Net;
using System.Threading;
using Aspose.Email.Mime;
using Aspose.Email.Mapi;
using Aspose.Email.Clients.Exchange.WebService;
using Aspose.Email.Clients.Exchange;

/*
This project uses Automatic Package Restore feature of NuGet to resolve Aspose.Email for .NET API reference 
when the project is build. Please check https://Docs.nuget.org/consume/nuget-faq for more information. 
If you do not wish to use NuGet, you can manually download Aspose.Email for .NET API from http://www.aspose.com/downloads, 
install it and then add its reference to this project. For any issues, questions or suggestions 
please feel free to contact us using http://www.aspose.com/community/forums/default.aspx
*/

namespace Aspose.Email.Examples.CSharp.Email.Exchange_EWS
{
    class ListFoldersFromExchangeServer
    {
        // ExStart:ListFoldersFromExchangeServer
        public static void Run()
        {
            try
            {
                IEWSClient client = EWSClient.GetEWSClient("https://outlook.office365.com/ews/exchange.asmx", "testUser", "pwd", "domain");
                Console.WriteLine("Downloading all messages from Inbox....");

                ExchangeMailboxInfo mailboxInfo = client.GetMailboxInfo();
                Console.WriteLine("Mailbox URI: " + mailboxInfo.MailboxUri);
                string rootUri = client.GetMailboxInfo().RootUri;
                // List all the folders from Exchange server
                ExchangeFolderInfoCollection folderInfoCollection = client.ListSubFolders(rootUri);
                foreach (ExchangeFolderInfo folderInfo in folderInfoCollection)
                {
                    // Call the recursive method to read messages and get sub-folders
                    ListSubFolders(client, folderInfo);
                }

                Console.WriteLine("All messages downloaded.");
            }
            catch (Exception ex)
            {
                Console.WriteLine(ex.Message);
            }
        }
        private static void ListSubFolders(IEWSClient client, ExchangeFolderInfo folderInfo)
        {
            // Create the folder in disk (same name as on IMAP server)
            Console.WriteLine(folderInfo.DisplayName);
            try
            {
                // If this folder has sub-folders, call this method recursively to get messages
                ExchangeFolderInfoCollection folderInfoCollection = client.ListSubFolders(folderInfo.Uri);
                foreach (ExchangeFolderInfo subfolderInfo in folderInfoCollection)
                {
                    ListSubFolders(client, subfolderInfo);
                }
            }
            catch (Exception)
            {
            }
        }
        // ExEnd:ListFoldersFromExchangeServer
    }
}